File: algorithm study/BOJ/2. Silver/1991.js
// boj 1911 트리 순회
// recursion

const input = require("fs").readFileSync("./example.txt").toString().split("\n");
// const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

const n = input[0].split(" ").map(Number); //1~26

let tree = {}; //객체: 노드와 자식의 값 배열을 하나로 저장
for (let i = 1; i <= n; i++) {
  const [node, left, right] = input[i].replace(/\r/gi, "").split(" "); //개행문자 \r 제거
  tree[node] = [left, right];
}
// console.log(tree);
// console.log(tree["A"]);

let ans = "";
function preorder(node) {
  if (node == ".") return;
  let [l, r] = tree[node];

  ans += node;
  preorder(tree[node][0]);
  preorder(tree[node][1]);
}
function inorder(node) {
  if (node == ".") return;
  let [l, r] = tree[node];

  inorder(tree[node][0]);
  ans += node;
  inorder(tree[node][1]);
}
function postorder(node) {
  if (node == ".") return;
  let [l, r] = tree[node];

  postorder(tree[node][0]);
  postorder(tree[node][1]);
  ans += node;
}

preorder("A");
ans += "\n";
inorder("A");
ans += "\n";
postorder("A");
ans += "\n";
console.log(ans);

File: algorithm study/BOJ/2. Silver/1991.cpp
#include <algorithm>
#include <iostream>
using namespace std;

int n;
char arr[26][3];
char ans[26][3];
void preorder(char node);
void inorder(char node);
void postorder(char node);

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    cin >> n;
    for (int i = 0; i < n; i++) cin >> arr[i][0] >> arr[i][1] >> arr[i][2];
    // for (int i = 0; i < n; i++) cout << arr[i][0] << " " << arr[i][1] << " " << arr[i][2] << "\n";
    preorder('A');
    for (int i = 0; i < n; i++) cout << arr[1][i];
    cout << "\n";
    inorder('A');
    postorder('A');
}

void preorder(char node) {
    if (node == '.') return;
    int cnt = 0;
    ans[0][cnt] = node;
    cnt++;

    // preorder(arr[find(arr, arr + n, node)][1]);
    preorder(arr[node][2]);
};
void inorder(char node){

};
void postorder(char node){};